services/builder/configProcessor.go

```diff
@@ -68,66 +49,115 @@ func configProcessor(c pb.ContextBoxServiceClient, wg *sync.WaitGroup) error {
+				}
+
+				if ctx, err = buildCluster(ctx, c); err != nil {
+					clusterView.SetWorkflowError(clusterName, err)
+					log.Error().Msgf("Failed to build cluster %s project %s : %v", clusterName, config.Name, err)
+					return err
+				}
+				log.Info().Msgf("First stage for cluster %s project %s finished building", clusterName, config.Name)
 
-			if done {
-				clusterView.SetWorkflowDone(clusterName)
-				log.Info().Msgf("Finished workflow for cluster %s project %s", clusterName, config.Name)
-				return updateWorkflowStateInDB(config.Name, clusterName, clusterView.ClusterWorkflows[clusterName], c)
+				// make the desired state of the temporary cluster the new current state.
+				clusterView.CurrentClusters[clusterName] = ctx.desiredCluster
+				clusterView.Loadbalancers[clusterName] = ctx.desiredLoadbalancers
 			}
 
-			// Handle deletion and addition of nodes.
-			tmpDesired, toDelete := stateDifference(clusterView.CurrentClusters[clusterName], clusterView.DesiredClusters[clusterName])
-			if tmpDesired != nil {
-				clusterView.ClusterWorkflows[clusterName].Description = "Processing stage [1/2]"
-				log.Info().Msgf("Processing stage [1/2] for cluster %s config %s", clusterName, config.Name)
+			if diff.ControlPlaneWithAPIEndpointReplace {
+				currentStage++
+				clusterView.ClusterWorkflows[clusterName].Description = fmt.Sprintf("Processing stage [%d/%d]", currentStage, stages)
+				log.Info().Msgf("Processing stage [%d/%d] for cluster %s config %s", currentStage, stages, clusterName, config.Name)
 
 				ctx := &BuilderContext{
+					projectName:    config.Name,
+					cluster:        clusterView.CurrentClusters[clusterName],
+					desiredCluster: clusterView.DesiredClusters[clusterName],
+					Workflow:       clusterView.ClusterWorkflows[clusterName],
+				}
+
+				if err := callUpdateAPIEndpoint(ctx, c); err != nil {
+					clusterView.SetWorkflowError(clusterName, err)
+					log.Error().Msgf("Failed to build cluster %s project %s : %v", clusterName, config.Name, err)
+					return err
+				}
+
+				clusterView.CurrentClusters[clusterName] = ctx.cluster
+				clusterView.DesiredClusters[clusterName] = ctx.desiredCluster
+				clusterView.DesiredLoadbalancers[clusterName] = ctx.desiredLoadbalancers
+
+				ctx = &BuilderContext{
 					projectName:          config.Name,
-					cluster:              clusterView.CurrentClusters[clusterName],
-					desiredCluster:       tmpDesired,
-					loadbalancers:        clusterView.Loadbalancers[clusterName],
-					desiredLoadbalancers: clusterView.DesiredLoadbalancers[clusterName],
-					deletedLoadBalancers: clusterView.DeletedLoadbalancers[clusterName],
+					desiredCluster:       clusterView.CurrentClusters[clusterName],
+					desiredLoadbalancers: clusterView.Loadbalancers[clusterName],
 					Workflow:             clusterView.ClusterWorkflows[clusterName],
 				}
 
-				if ctx, err = buildCluster(ctx, c); err != nil {
+				if err := callKubeEleven(ctx, c); err != nil {
```

Comment:
Why are we skipping terraformer/ansibler/kuber here?

Reply:
The idea is a follows.

if you have the following config:

```yaml
kubernetes:
  clusters:
    - name: ts4-c-1
      version: v1.24.0
      network: 192.168.2.0/24
      pools:
        control:
          - hetzner-control
        compute:
          - hetzner-compute

```

and then change it to (as an example): 

```yaml
kubernetes:
  clusters:
    - name: ts4-c-1
      version: v1.24.0
      network: 192.168.2.0/24
      pools:
        control:
          - gcp-control
        compute:
          - hetzner-compute
          - gcp-compute
loadBalancers:
  roles:
    - name: apiserver-lb
      protocol: tcp
      port: 6443
      targetPort: 6443
      target: k8sControlPlane
  clusters:
    - name: ts4-lb-1
      roles:
        - apiserver-lb
      dns:
        dnsZone: claudie.dev
        provider: cloudflare-1
      targetedK8s: ts4-c-1
      pools:
        - gcp-lb
```


In the first stage here:
https://github.com/berops/claudie/blob/f9fbe0ed03f641356c01283da3638b8363d79ef2/services/builder/configProcessor.go#L63-L90

It will run the changes but without LBs as there is a clear dependency issue (you can attach a LB and change the api endpoint as kubelet and other binaries aren't even installed until the kube-eleven stage)


After that point you'll have a state that resembles something like this (note no lbs build yet):
```yaml
kubernetes:
  clusters:
    - name: ts4-c-1
      version: v1.24.0
      network: 192.168.2.0/24
      pools:
        control:
          - hetzner-control
          - gcp-control
        compute:
          - hetzner-compute
          - gcp-compute
```

So all the nodes have been build and went through the workflow and when we enter the stage
https://github.com/berops/claudie/blob/f9fbe0ed03f641356c01283da3638b8363d79ef2/services/builder/configProcessor.go#L93

we basically work with the in-between state so we just transfer the api endpoint to the new control plane, run kube-eleven to get the kubeconfig and patch it in kuber. No need to run ansibler terraformer etc... (has been done)

Then the deletion of nodes follows and building the state now with also the Lbs.